### PR TITLE
add graylog compatibility

### PR DESCRIPTION
**Pull request by negbie**

### CODE REVIEW

File: data/kamailio.cfg [hunk 1]
```diff
@@ -1268,9 +1552,13 @@ route[SEND_ES_STATS] {
 	$http_req(hdr) = "Content-Type: application/json; charset=UTF-8";
 
 	# Index+Type defined in bulk object
-	$var(elasticurl) = "ELASTICSEARCH_URL/_bulk";
+	$var(elasticurl) = "http://YOUR_GRAYLOG_IP:12201/gelf";
```
Comment (lmangani) [inline data/kamailio.cfg]: should $var(elasticurl) also be conditional and retain support both GRAYLOG and existing ES?

Comment (negbie) [reply data/kamailio.cfg]: definitely

File: data/kamailio.cfg [hunk 2]
```diff
@@ -1267,10 +1552,14 @@ route[SEND_ES_STATS] {
...
 	$http_req(hdr) = "Content-Type: application/json; charset=UTF-8";
 
+	#!ifdef TO_GRAYLOG
+	$var(graylogurl) = "http://YOUR_GRAYLOG_IP:12201/gelf";
+	http_async_query("$var(graylogurl)", "$var(graylog_json)", "ES_REPLY");
+	#else
 	# Index+Type defined in bulk object
-	$var(elasticurl) = "ELASTICSEARCH_URL/_bulk";
-
+	$var(elasticurl) = "http://YOUR_ELASTIC_IP:9200/_bulk";
```
Comment (grumvalski) [inline data/kamailio.cfg]: In this way you are hardcoding the protocol and the port on which elasticsearch (or graylog) run. I still think here is better to have the full url which could be composed elsewhere by configuration variables (e.g.: transport, ip, port).

Comment (negbie) [reply data/kamailio.cfg]: Good point. Will change that!

File: data/kamailio.cfg [hunk 3]
```diff
@@ -586,17 +770,29 @@ route[CHECK_ALARM]
...
 						$var(alarm_data) = $var(alarm_data) + " {\"index\": {\"_index\":\"homer-"+ $timef(%d.%m.%Y) +"\",\"_type\":\""+$var(type)+"\" }}\n";
 						$var(alarm_data) = $var(alarm_data) + $var(tmp_json) + "\n";
 
+						
+						#!ifdef TO_GRAYLOG
+						$var(short) = @cfg_get.short.message;
+						$var(full) = $sel(msg);
+						jansson_set("string", "short_message", ""+$var(short)+"", "$var(tmp_json)");
+						jansson_set("string", "full_message", ""+$var(full)+"", "$var(tmp_json)");
+						$var(graylog_json) = $var(tmp_json);
```
Comment (grumvalski) [inline data/kamailio.cfg:831]: This part looks to me repeated in the script. Wouldn't it be better to put it in the SEND_ES_DATA route inside the 
#!ifdef TO_GRAYLOG?
So, using jansson_set on the $var(body).

File: data/kamailio.cfg [hunk 4]
```diff
@@ -768,8 +984,20 @@ route[CHECK_STATS] {
...
+		jansson_set("string", "short_message", ""+$var(short)+"", "$var(tmp_json)");
+		jansson_set("string", "full_message", ""+$var(full)+"", "$var(tmp_json)");
+		$var(graylog_json) = $var(tmp_json);
+		#!endif
+		
 		$var(tmp_json) = 0;
 		$sht(a=>packet::count) = 0;
+
+		$var(body) = $var(stats_data);
+		route(SEND_ES_DATA);
```
Comment (grumvalski) [inline data/kamailio.cfg:1030]: Why did you split data sending in this way. In this way you are basically sending an HTTP request per document (elasticsearch or graylog), loosing the benefit of the _bulk API. For this reason the stats where grouped by type.

File: data/kamailio.cfg [hunk 5]
```diff
@@ -353,129 +364,289 @@ onreply_route {
...
+			$var(tmp_graylog_json) = 0;
+
+			if($sht(x=>$ci::callid_count) == $null) $sht(x=>$ci::callid_count) = 0;
+			$sht(x=>$ci::callid_count) = $sht(x=>$ci::callid_count) + 1;
+
+			route(SEND_ES_DATA);
+		}
+	}
+}
+
```
Comment (grumvalski) [inline data/kamailio.cfg:678]: Really nice one. I'd like to have the x-rtp stats on elasticsearch too. Could we generalize this part in order to follow the same logic as per the other stats (so valid both for elasticsearch and graylog).

**Review by grumvalski — COMMENTED** [review]
Finally I could have a look on this PR. Overall good, some remarks.

File: data/kamailio.cfg (hunk 5 above)
Comment (negbie) [reply data/kamailio.cfg:678]: will add it.

File: data/kamailio.cfg (hunk 3 above)
Comment (negbie) [reply data/kamailio.cfg:831]: I am still not confident with my field naming. the full_message field is optional and maybe i will remove it completly. short_message is mandatory for gelf. if i decide to keep it static i will move it to SEND_ES_DATA.

File: data/kamailio.cfg (hunk 4 above)
Comment (negbie) [reply data/kamailio.cfg:1030]: For graylog you cannot send directly to es. You need to spawn and gelf http input and i dont know if this input type supports the _bulk api. I need to google it first or ask in the user group.

Comment (negbie) [reply data/kamailio.cfg:1030]: But nevertheless i will change it to get a similar behaviour to elastic

File: data/kamailio.cfg [hunk 6]
```diff
@@ -537,113 +538,121 @@ route[PARSE_X_RTP_STATS] {
...
+		$var(graylog_json) = $var(tmp_json);
+		#!endif
+		
+		$var(tmp_json) = 0;
+
+		if($sht(x=>$ci::callid_count) == $null) $sht(x=>$ci::callid_count) = 0;
+		$sht(x=>$ci::callid_count) = $sht(x=>$ci::callid_count) + 1;
+		
+		$var(body) = $var(xrtp_stats);
+		route(SEND_ES_DATA);
```
Comment (grumvalski) [inline data/kamailio.cfg]: thanks :)

**Comment by lmangani** [comment]
Awesome job guys! Are we looking good for a merge or shall we modularize this further?

**Comment by negbie** [comment]
Lets wait a bit till i get a final answer from here:
https://groups.google.com/forum/#!topic/graylog2/nTr8vFdlizc
Still then i need to polish a bit ;)

**Comment by grumvalski** [comment]
two points still to address imho:
1) code duplication for graylog logs
2) bulk/aggregated requests

File: data/kamailio.cfg (hunk 6 above)
Comment (negbie) [reply data/kamailio.cfg]: Your welcome i still need to add but will do in another pr

jansson_set("string", "host", "$var(host)", "$var(tmp_json)");
jansson_set("string", "timestamp", "$var(isotime)", "$var(tmp_json)");

**Comment by negbie** [comment]
1) true but maybe i will let it inside the #!ifdef TO_GRAYLOG. I mean it does not hurt there anybody and i will need that ifdef anyway because i need another timestamp for graylog "timestamp", ""+$TS+"."+$TV(u)/999+""
One advantage is when you need additional information for example you want the whole message body for 50X response codes you can easily add it to short_message or full_message without add extra fields. In graylog you can easily extract those information.

2) it seems graylog cannot handle it (yet) in the same way like sending it directly to elastic. So i will need route(SEND_ES_DATA) for every request but i will move it to the ifdef so it will affect only graylog users.

**Comment by grumvalski** [comment]
Understand. 
We'll think about refactoring and ave more compact code later :)

**Comment by negbie** [comment]
Btw. i noticed that we we need to take care about the private memory pool space. run sometimes into "not enough memory to initialize destination branches" errors.

Maybe we should raise PKG_MEMORY in /etc/default/kamailio to 16 or make ifdefs for elasticsearch and graylog.

**Comment by grumvalski** [comment]
This is strange because I don't think that we are using so much the package memory, most of the structure being in shm. Do you see which child is printing the rror: maybe the http async worker(s)? When this occur is it something that stay "stable" or the errors are only from time to time? could you run
kamcmd pkg.stats
kamcmd mod.stats all pkg

when this happens?

**Comment by negbie** [comment]
Ok it seems there is no memory leak but just not enough memory during the startup of kamailio. When i remove some private memory variables till kamailio is able to start you have more or less those stats for all the workers:

{
        entry: 10
        pid: 19936
        rank: -2
        used: 4542944
        free: 2951136
        real_used: 5437472
        total_size: 8388608
        total_frags: 1464
}

**Comment by negbie** [comment]
It comes from the regular expressions which i use a lot in route[PARSE_X_RTP_STATS]

File: data/kamailio.cfg [hunk 7]
```diff
@@ -1306,87 +2086,100 @@ route[CHECK_STATS] {
...
 
-	if ($var(body) == "") return;
-
 	$http_req(suspend) = 0;
 	$http_req(method) = "POST";
 	$http_req(hdr) = "Content-Type: application/json; charset=UTF-8";
 
-	# Index+Type defined in bulk object
+#!ifdef TO_GRAYLOG
+	if ($var(graylog_json) == $null || $var(graylog_json) == 0 ) return;
```
Comment (grumvalski) [inline data/kamailio.cfg]: a $var cannot be $null in kamailio, either is "" or 0.

File: data/kamailio.cfg [hunk 8]
```diff
@@ -1306,87 +2086,100 @@ route[CHECK_STATS] {
...
 	sht_rm_name_re("d=>.*");
-	
-	$var(body) = $var(stats_generic);
 
+	$var(body) = $var(stats_generic);
 	route(SEND_ES_DATA);
 #!endif
 }
 
 route[SEND_ES_DATA] {
```
Comment (grumvalski) [inline data/kamailio.cfg:2114]: maybe we could rename this route to something like SEND_HTTP_DATA (and the reply accordingly).

**Comment by grumvalski** [comment]
Looks good. I think at this point we should merge, what do you think @lmangani ?

File: data/kamailio.cfg (hunk 8 above)
Comment (negbie) [reply data/kamailio.cfg:2114]: makes sense.

**Comment by grumvalski** [comment]
let's merge it!

**Pull request merged**